flocker/node/agents/pd.py

```diff
@@ -0,0 +1,371 @@
+
+    :returns: The `unicode` name of the instance the disk is attached to or
+        `None` if it is not attached to anything.
+    """
+    # TODO(mewert): determine how this works with a disk being attached to
+    # multiple machines, update comment above.
+    users = disk.get('users', [])
+    if not users:
+        return None
+    return unicode(users[0].split('/')[-1])
+
+
+@implementer(IBlockDeviceAPI)
+class PDBlockDeviceAPI(object):
+    """
+    A PD implementation of ``IBlockDeviceAPI`` which creates block devices in a
+    GCE project.
+
+    Constraints imposed from GCE:
+        - GCE does not have a grab-bag of metadata you can attach to disks.
+            Instead, it has two fields in the disk resource you can set:
+                - name (must be unique within a project,
+                        limited to 63 characters,
+                        must start with letter)
+                - description (free form text field)
+        - GCE does allow you to pick the blockdevice_id of your volumes (the
+            unique identifier that will be used to identify your volume on
+            subsequent API calls). In GCE terms this is the disk resource name.
+        - GCE lets you filter on both of these string fields, but only for
+            equals and not equals.
+        - GCE lets you set a token in the SCSI info pages on your blockdevices
+            every time you attach them to a VM so that symlinks are created at
+            '/dev/disk/by-id/google-<token>'.
+            - This technically is brittle to implementation, configuration, and
+                version of udev and the SCSI driver within the guest OS.
+        - GCE resource names and descriptions are immutable after creation.
+
+    Design:
+        - For a given dataset_id (06e07bcc-fea1-4810-8c6d-7487196998b6) prefix
+            the dataset_id with "flocker-v1-" to get the unicode blockdevice-id
+            (flocker-v1-06e07bcc-fea1-4810-8c6d-7487196998b6).
+        - Set the name of the GCE disk resource to the blockdevice_id. This
+            lets us perform operations on the disk only knowing its
+            blockdevice_id.
+        - Set the description of the disk resource to:
+            "flocker-v1-cluster: <cluster-uuid>".
+        - Whenever attaching a disk to an instance, set <token> to the
+            blockdevice_id.
+
+    Design implications:
+        - The GCE disk names meet the length and format requirements, and are
+            as unique as blockdevice_ids.
+        - We can perform operations on a disk only knowing its blockdevice_id
+            without making additional API calls or looking things up in a
+            table.
+        - dataset_id is a pure function of blockdevice_id and vice versa.
+        - You can have multiple clusters within the same project.
+        - Multiple clusters within the same project cannot have datasets with
+            the same UUID. **
```

Comment:
UUIDs are supposed to be universal, and we assume that in lots of places, so that seems fine.

Except... users can pick dataset IDs! We ripped out logic from Docker plugin that requires that as of next release, so there's just `flocker-deploy` (which is a toy, and will be deprecated soon I hope) and the fact our API allows it. But really if you're picking duplicate IDs you're doing it wrong, so I think it's fine to have this restriction so long as it's documented.